### PR TITLE
[torch.compile][TPU] Make @support_torch_compile work for XLA backend

### PR DESCRIPTION
**Pull request by lsy323**
Make `@support_torch_compile` work for XLA backend. With the custom dispatcher, overhead of dynamo guard evaluation is eliminated.

For TPU backend, each models have 2 FX graphs/dynamo bytecodes:
1. During profiling run - No KV cache
2. After profiling run - with KV cache

This breaks the assumption in the current `@support_torch_compile` implementation - Each model has one FX graph/cached bytecode. Since the profiling graph won't be invoked after the profiling run, we clear the cached bytecode after profiling run so that the assumption remains true.

Other changes:
- Remove `ModelWrapperV1`, which is used to wrap the model code and `torch.compile` the wrapped model. It's not needed anymore since we are reusing the compile decorator.
- Since `ModelWrapperV1` is removed, sampler logic is moved to a separate function.

Credit to @WoosukKwon on the idea of clearing bytecode cache and @youkaichao for the many help on torch dynamo related questions!

cc @youkaichao @alexm-redhat @miladm @NickLucche @WoosukKwon @yaochengji @robertgshaw2-redhat

### CODE REVIEW

**Comment by lsy323** [comment]
Slightly improved the throughput `6.09` -> `6.14 req/s`, benchmarking cmd is:
```
VLLM_USE_V1=1 vllm serve meta-llama/Llama-3.1-8B-Instruct \
 --disable-log-requests \
 --port 8004 \
 --gpu-memory-utilization 0.95 \
 --max-num-seqs 512 \
 --max-num-batched-tokens 512 \
 --tensor-parallel-size 1 \
 --max-model-len 2048

python benchmarks/benchmark_serving.py \
    --backend vllm \
    --model meta-llama/Llama-3.1-8B-Instruct  \
    --dataset-name random \
    --random-input-len 1800 \
    --random-output-len 128 \
    --random-prefix-len 0 \
    --port 8004
```

File: vllm/v1/worker/tpu_model_runner.py [hunk 1]
```diff
@@ -772,11 +770,9 @@ def _dummy_run(self, kv_caches, num_tokens: int) -> None:
...
 
         with set_forward_context(attn_metadata, self.vllm_config, 0):
-            out = self.model(input_ids=input_ids,
-                             positions=position_ids,
-                             kv_caches=kv_caches,
-                             inputs_embeds=inputs_embeds)
-        self._hidden_states_dtype = out.dtype
+            self.model(input_ids=input_ids,
+                       positions=position_ids,
+                       inputs_embeds=inputs_embeds)
```
Comment (mgoin) [inline vllm/v1/worker/tpu_model_runner.py:775]: Please keep the _hidden_states_dtype assignment

File: vllm/v1/worker/tpu_model_runner.py [hunk 2]
```diff
@@ -786,9 +782,9 @@ def capture_model(self) -> None:
         start = time.perf_counter()
         for num_tokens in self.num_tokens_paddings:
             logger.info("  -- num_tokens: %d", num_tokens)
-            self._dummy_run(self.kv_caches, num_tokens)
+            self._dummy_run(num_tokens)
             xm.mark_step()
-        xm.wait_device_ops()
+            xm.wait_device_ops()
```
Comment (mgoin) [inline vllm/v1/worker/tpu_model_runner.py:787]: Is there a reason to pull the device sync inside of the loop? IIRC we pulled it out since it made parallel compilation slightly quicker

File: vllm/v1/worker/tpu_model_runner.py [hunk 3]
```diff
@@ -874,79 +869,45 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
...
-                       hidden_states: torch.Tensor) -> Optional[torch.Tensor]:
-        # SamplingMetadata here for pruning output in LogitsProcessor, disabled
-        logits = self.model.compute_logits(hidden_states, None)
-        return logits
-
-    def get_multimodal_embeddings(self, *args, **kwargs):
-        return self.model.get_multimodal_embeddings(*args, **kwargs)
-
-    def get_input_embeddings(self, *args, **kwargs):
-        return self.model.get_input_embeddings(*args, **kwargs)
```
Comment (mgoin) [inline vllm/v1/worker/tpu_model_runner.py:948]: Have you tested that multimodal inference still works and these are called correctly?

File: vllm/v1/worker/tpu_model_runner.py [hunk 4]
```diff
@@ -874,79 +869,45 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
...
-
-        hidden_states = self.model(
-            input_ids=input_ids,
-            positions=positions,
-            inputs_embeds=inputs_embeds,
-        )
-
-        return hidden_states
+    def reset_dynamo_cache(self):
+        # TODO(lsy323): Support multimodal models, the backbone language model
```
Comment (NickLucche) [inline vllm/v1/worker/tpu_model_runner.py:873]: do you mean 
```suggestion
        compiled_model = self.model.language_model if self.is_multimodal_model else self.model.model
        
```

File: vllm/v1/worker/tpu_model_runner.py [hunk 5]
```diff
@@ -874,79 +869,45 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
             self.vllm_config.compilation_config.static_forward_context,
             self.kv_caches)
 
-
-class ModelWrapperV1(nn.Module):
-
-    def __init__(self, model: nn.Module):
```
Comment (NickLucche) [inline vllm/v1/worker/tpu_model_runner.py:880]: I didn't mind the wrapper too much I think it was grouping a few related functions nicely. Still, if it benefits performance I am ok with that.

File: vllm/v1/worker/tpu_model_runner.py [hunk 6]
```diff
@@ -874,79 +869,45 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
...
-        separately from `forward` for lighter compilation overhead.
-        """
+            Sample with xla-friendly function. This function is to be traced 
+            separately for lighter compilation overhead.
+            """
         # Tensor `sample_hidden_states` is of fixed pre-compiled size.
         sample_hidden_states = \
             hidden_states[sampling_metadata.indices_do_sample]
-        logits = self.compute_logits(sample_hidden_states)
+        logits = self.model.compute_logits(sample_hidden_states, None)
```
Comment (NickLucche) [inline vllm/v1/worker/tpu_model_runner.py:894]: can we re-add the pruning comment here? Just in case it slips though in the future.

File: vllm/v1/worker/tpu_model_runner.py (hunk 1 above)
Comment (NickLucche) [reply vllm/v1/worker/tpu_model_runner.py:775]: yep that's still needed

**Review by NickLucche — CHANGES_REQUESTED** [review]
Thanks for the work!
Left a comment about MM, I think that's about the only thing to clarify on my side.

File: vllm/v1/worker/tpu_model_runner.py (hunk 3 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:948]: I only ran TPU CI to test the change, multimodal is not tested. Can you provide a script to test the multimodal? I can test it on this PR.

File: vllm/v1/worker/tpu_model_runner.py (hunk 1 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:775]: Here is more context https://github.com/vllm-project/vllm/pull/15714/files#r2019453889

File: vllm/v1/worker/tpu_model_runner.py (hunk 4 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:873]: I haven't tested multimodal yet, plan to do it in another PR. Right now I only tested the tests in TPU CI

File: vllm/v1/worker/tpu_model_runner.py (hunk 5 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:880]: The wrapper was added in V0 for `torch.compile`. Now we have the `@support_torch_compile` decorator that wraps the model with `torch.compile` already, therefore we don't need the wrapper anymore.

File: vllm/v1/worker/tpu_model_runner.py (hunk 6 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:894]: Sorry, may I ask what is 'pruning comment'?

File: vllm/v1/worker/tpu_model_runner.py (hunk 1 above)
Comment (NickLucche) [reply vllm/v1/worker/tpu_model_runner.py:775]: I think this is out of scope, but yes in principle you'd only need `self.dtype`. The dtype issue you linked has proven that nothing crashes at runtime should the output of an op not match `self.dtype` though.
Hence if the same bug were to re-appear again, we would only notice the server recompiling and we'd have to debug again the way I did, painfully.

File: vllm/v1/worker/tpu_model_runner.py (hunk 3 above)
Comment (NickLucche) [reply vllm/v1/worker/tpu_model_runner.py:948]: `VLLM_USE_V1=1 vllm serve llava-hf/llava-1.5-7b-hf --max-model-len 4096 --max-num-seqs 8 --max-num-batched-tokens 512 --chat-template examples/template_llava.jinja`

then `python examples/online_serving/openai_chat_completion_client_for_multimodal.py`

File: vllm/v1/worker/tpu_model_runner.py (hunk 4 above)
Comment (NickLucche) [reply vllm/v1/worker/tpu_model_runner.py:873]: > I haven't tested multimodal yet, plan to do it in another PR

I think it's best if we double check here o/w we may inadvertently break MM

File: vllm/v1/worker/tpu_model_runner.py (hunk 6 above)
Comment (NickLucche) [reply vllm/v1/worker/tpu_model_runner.py:894]: `# SamplingMetadata here for pruning output in LogitsProcessor, disabled`
or smt along this line to indicate why the 2nd argument is None. That's because it could enable logits pruning which is bad for xla

File: vllm/v1/worker/tpu_model_runner.py (hunk 3 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:948]: sg I'll rebase to your PR after :)

File: vllm/v1/worker/tpu_model_runner.py (hunk 6 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:896]: Thanks! Updated

File: vllm/v1/worker/tpu_model_runner.py (hunk 4 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:874]: I'll rebase to the MM PR after it's merged to ensure this PR doesn't break MM

File: vllm/v1/worker/tpu_model_runner.py (hunk 2 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py]: Thanks! Updated.

File: vllm/v1/worker/tpu_model_runner.py (hunk 1 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py]: Thanks! Updated

File: vllm/v1/worker/tpu_model_runner.py [hunk 7]
```diff
@@ -874,79 +870,46 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
...
-            input_ids=input_ids,
-            positions=positions,
-            inputs_embeds=inputs_embeds,
-        )
-
-        return hidden_states
+    def reset_dynamo_cache(self):
+        # TODO(lsy323): Support multimodal models, the backbone language model
+        # is stored in a different member.
+        compiled_model = self.model.model
```
Comment (vanbasten23) [inline vllm/v1/worker/tpu_model_runner.py:876]: i wonder why we need to do `model.model`. Could you add a comment?

Comment (yaochengji) [reply vllm/v1/worker/tpu_model_runner.py:876]: Is it possible `self.model` doesn't have a attr `model`? E.g. it's not annotated by `support_torch_compile`?

Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:876]: Yes, `model.model` is torch compiled

File: vllm/v1/worker/tpu_model_runner.py (hunk 3 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:948]: Hi @NickLucche Thank you for providing the testing cmd for multimodal, I made this PR working for `llava-hf/llava-1.5-7b-hf`. The server cmd runs fine.

However, the client script `python examples/online_serving/openai_chat_completion_client_for_multimodal.py` would fail at HEAD.

File: vllm/v1/worker/tpu_model_runner.py [hunk 8]
```diff
@@ -874,79 +870,47 @@ def initialize_kv_cache(self, kv_cache_config: KVCacheConfig) -> None:
...
-            positions=positions,
-            inputs_embeds=inputs_embeds,
-        )
-
-        return hidden_states
+    def reset_dynamo_cache(self):
+        if self.is_multimodal_model:
+            compiled_model = self.model.language_model.model
+        else:
+            compiled_model = self.model.model
```
Comment (mgoin) [inline vllm/v1/worker/tpu_model_runner.py:877]: AFAIK "language_model" is not a stable attribute to reference, it is based on the HF model definition. Maybe @ywang96 @DarkLight1337 would know a stable interface to access the language model backbone?

Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:877]: I think a potential solution is to add a query function `get_backbone_lm` for multimodal models, similar to the existing `get_input_embeddings`

File: vllm/v1/worker/tpu_model_runner.py (hunk 3 above)
Comment (lsy323) [reply vllm/v1/worker/tpu_model_runner.py:948]: @NickLucche Thank you so much for trying! I set up a new conda env on my end and tried as well. I also found it's really slow, was about to ask you lol.

File: vllm/v1/worker/tpu_model_runner.py (hunk 8 above)
Comment (DarkLight1337) [reply vllm/v1/worker/tpu_model_runner.py:877]: Yeah we can add `get_language_model` to `SupportsMultiModal` interface

**Review by yaochengji — APPROVED** [review]
LGTM, thanks!

File: vllm/v1/worker/tpu_model_runner.py (hunk 8 above)
Comment (NickLucche) [reply vllm/v1/worker/tpu_model_runner.py:877]: Something like https://github.com/vllm-project/vllm/pull/16007?

Comment (mgoin) [reply vllm/v1/worker/tpu_model_runner.py:877]: Please leave an assert here for now as we will address later @lsy323

**Review by mgoin — APPROVED** [review]
For now I think we should aim to land this just working with language_model since there are other blockers/developments there. Afterwards we can migrate to use the get_language_model interface proposed by Nicolo.
